Add window.level to set window level (Normal, AlwaysOnTop).

davehorner opened the pull request:
Great software.  Thank you.

davehorner commented on alacritty/src/display/window.rs:
```diff
@@ -298,7 +299,8 @@ impl Window {
     #[cfg(windows)]
     pub fn get_platform_window(_: &Identity, window_config: &WindowConfig) -> WindowAttributes {
-        let icon = winit::window::Icon::from_resource(IDI_ICON, None);
+        let icon: std::result::Result<winit::window::Icon, winit::window::BadIcon> =
```
this change is made by vscode automatically...very fast without interruption it might be worth going to not have to deal with it.

davehorner replied:
normal is the default.

tested using local git:
```
cargo run -- -o window.window_level="'AlwaysOnTop'"
```
on top.

I tried
```
cargo run -- -o window.window_level="'AlwaysOnBottom'"
```
and it isn't starting behind the windows.  it did when I was testing via a direct function set_window_level that I removed as deadcode.  I'm not sure what's up there but normal and always on top are working great for me.

davehorner replied:
ubuntu 24.10 doesn't seem to apply the setting...command line parsed into three options, AlwaysOnTop doesn't keep it on top with given code.

nixpulvis replied:
Are the explicit type annotations even needed? Isn't the return type of `from_resource` enough?

davehorner replied:
that I left in as VS code is doing that automatically.  it can be pulled out; I don't know why vscode feels so strongly about that line.

chrisduerr replied:
This should be removed.

nixpulvis commented:
FWIW, I could see using this, though it sounds like it's not supported on Wayland which is unfortunate.

This is another example of a feature that could probably be solved by the window manager as well. But we have configurations for things like the window position and default window mode, so I think this fits in pretty naturally.

chrisduerr commented:
> But we have configurations for things like the window position and default window mode, so I think this fits in pretty naturally.

Window positioning controls were a mistake and would not exist in Alacritty if it was written today.

nixpulvis commented:
> Window positioning controls were a mistake and would not exist in Alacritty if it was written today.

Yea I agree with that.

But I still think this fits in somewhat well with window mode, opacity, etc.

davehorner commented:
> Why do you want this in Alacritty? Why are you adding things that you confirmed yourself to not be working?

I understand some people don't want this stuff; others do.  I made a contribution.  It's not finished.  I guess I just have to come with everything finished?  I saw a thing I wanted to do and I did it.

nixpulvis commented:
> I guess I just have to come with everything finished?

Personally, I like the GitHub draft PRs for things like this. That way it's clear it's not done yet.

davehorner commented:
my apologies,  I can pull it if I'm out of place with an draft change.  I've not done "draft" PRs before.

chrisduerr commented:
Feel free to mark it as ready for review once you think it's ready. If you don't plan on getting it to the finish line, please close the PR.

davehorner commented:
Let me know what you think about expanding it to an WindowLevelStrategy or WindowLevelFocus or a different option entirely or....or maybe winit#1944 not a good idea.  maybe too much for those who dislike these sorts of concerns being in their terminal.

its not only about ontop but also about focus control.  I will try and figure out the PR problems this weekend.

davehorner commented:
I spent a few hours on it last night and a few hours this morning trying to get AlwaysOnBottom to work.  I am not sure what is going on.  I modified winit locally to debug and I am seeing the proper constants being set.

```
running cargo run -- -o window.window_level="'Normal'"
winit::window::WindowAttributes::with_window_level Normal
winit::window::init TOP false
winit::window::init BOTTOM false
on taskbar
winit::set_window_level level=TOP false
winit::set_window_level level=BOTTOM false
```
```
cargo run -- -o window.window_level="'AlwaysOnTop'"
winit::window::WindowAttributes::with_window_level AlwaysOnTop
winit::window::init TOP true
winit::window::init BOTTOM false
on taskbar
winit::WindowFlags::diff.intersects insects WindowFlags::ALWAYS_ON_TOP(true) or WindowFlags::ALWAYS_ON_BOTTOM(false)
HWND_TOPMOST
winit::set_window_level level=TOP true
winit::set_window_level level=BOTTOM false
```
```
cargo run -- -o window.window_level="'AlwaysOnBottom'"
self.window_level.unwrap_or_default()=AlwaysOnBottom
winit::window::WindowAttributes::with_window_level AlwaysOnBottom
self.window_level.unwrap_or_default()=AlwaysOnBottom
winit::window::WindowAttributes::with_window_level AlwaysOnBottom
winit::window::init TOP false
winit::window::init BOTTOM true
on taskbar
winit::WindowFlags::diff.intersects insects WindowFlags::ALWAYS_ON_TOP(false) or WindowFlags::ALWAYS_ON_BOTTOM(true)
HWND_BOTTOM
winit::set_window_level level=TOP false
winit::set_window_level level=BOTTOM true
```

I'm stumped and I'm actually not interested in this feature in particular.  I removed the enum for AlwaysOnBottom.

This PR can be merged for Normal and AlwaysOnTop support.
If the lack of OnBottom means it doesn't go; that's fine.  I had fun.

davehorner commented:
i don't own a mac to test; I will look for a friend, or let me know what you find if you've got one.

chrisduerr commented on alacritty/src/config/window.rs:
```diff
@@ -61,6 +61,9 @@ pub struct WindowConfig {
     /// System decorations theme variant.
     decorations_theme_variant: Option<Theme>,
+
+    /// Window level.
+    pub window_level: Option<WindowLevel>,
```
What is the difference between `None` and `WindowLevel::Normal`? Either `WindowLevel::Normal` should be removed, or the option. Otherwise this seems redundant.

chrisduerr commented on alacritty/src/config/window.rs:
```diff
@@ -306,3 +315,20 @@ impl From<Theme> for WinitTheme {
     }
 }
+
+/// WindowLevel.
```
This doc comment is just the enum name, making it completely useless.

chrisduerr commented on alacritty/src/config/window.rs:
```diff
@@ -6,7 +6,7 @@ use serde::{Deserialize, Deserializer, Serialize};
 #[cfg(target_os = "macos")]
 use winit::platform::macos::OptionAsAlt as WinitOptionAsAlt;
-use winit::window::{Fullscreen, Theme as WinitTheme};
+use winit::window::{Fullscreen, Theme as WinitTheme, WindowLevel as WinitWindowLevel};
```
Why was this renamed to `WinitWindowLevel`?

chrisduerr reviewed (changes requested):
Missing a changelog entry.

kchibisov commented on alacritty/src/config/window.rs:
```diff
@@ -61,6 +61,9 @@ pub struct WindowConfig {
     decorations_theme_variant: Option<Theme>,
+
+    /// Window level.
+    pub window_level: WindowLevel,
```
This should be called just `level`, since it's in `window` section already.

kchibisov commented on CHANGELOG.md:
```diff
@@ -10,6 +10,14 @@ Notable changes to the `alacritty_terminal` crate are documented in its
 
+### Added
+
+- AlwaysOnTop window level provides always on top window placement on platforms that support it.
```
This should document the new config option, like `window.level` to set preferred window level.

kchibisov commented on CHANGELOG.md:
```diff
@@ -10,6 +10,14 @@ Notable changes to the `alacritty_terminal` crate are documented in its
+
+- AlwaysOnTop window level provides always on top window placement on platforms that support it.
+
+### Fixed
+
+- Mouse/Vi cursor hint highlighting broken on the terminal cursor line
```
This looks like a rebase artifact.

kchibisov commented on alacritty/src/config/window.rs:
```diff
@@ -306,3 +315,19 @@ impl From<Theme> for WinitTheme {
+#[derive(ConfigDeserialize, Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WindowLevel {
+    #[default]
+    Normal,
```
Add `AlwaysOnBottom`.

kchibisov reviewed (dismissed):
You should document your config option inside the `alacritty.5.scd`.

davehorner commented:
The PR I'm submitting doesn't include toggling in realtime; I'd like to see that in the future. 

> [rust-windowing/winit#1944](https://github.com/rust-windowing/winit/pull/1944) -timing ontop and not is really where we end up and its not pretty. then people want to toggle in realtime.
> 

I see we just changed focus behavior on mac.  
https://github.com/alacritty/alacritty/commit/39ea7271e32ad88280191d8040d6f8feafe4307a
``
        // Always focus new windows, even if no Alacritty window is currently focused.
        #[cfg(target_os = "macos")]
        window.focus_window();
```
Why is this only for MacOs?  I think this should be something under config option?

kchibisov commented:
>Why is this only for MacOs? I think this should be something under config option?

other platforms behave sane here, so they don't need such thing.

davehorner commented:
> other platforms behave sane here, so they don't need such thing.

stealing focus might be an OK default on mac and not needed on other platforms.
Might it be nice to have an option to steal focus when desired on those platforms that are sane?
Or to allow for not focus'ing at all.

This focus_window() may cause issues with AlwaysOnBottom.  I'm not sure if focusing brings to the top or not.
It would be interesting to see a window start on bottom and focus while still under all the other windows.

kchibisov commented:
If you open a window yourself, it should be focused by default. On macOS it's not the case in some cases, so you have to find where this window got opened.

The main problem is that _you_ do it, and you don't have any idea where the window is, since macOS thinks that it shouldn't be focused.

davehorner commented:
I see complicated and there's all sorts of history...
https://github.com/rust-windowing/winit/pull/1764 - request_user_attention 

> "un-requesting" user attention seems like a rare use case

start ontop without focus, or ontop with focus for a period and drop back to HWND_BOTTOM placing focus back to source.

as you said, open a window and focus is given by default.  tracking window focus source to "un-request" user attention isn't easy; it seems there's been some thought into it.  Hard coding a set_focus() on mac would make ontop without focus not possible if I understand correctly.

kchibisov commented:
`set_focus` is only called once you built a window, once you focus something else, it'll lose focus. It's only a matter whether you have initial focus or not here on macOS when you run it in some weird way. I really don't see what's the problem is here and what breaks, adding options for every setting is not an option though.

User attention has nothing to do with focus or anything, it's an indicator that something happened in the app and it should be focused.

davehorner commented:
I think limiting the change to AlwaysOnTop reduces some of my concerns about focus and placement interacting.
I can add AlwaysOnBottom back; it adds some complexity and its not working now.

davehorner commented:
I have a mac now.  Starting ```cargo run -- -o window.level="'AlwaysOnBottom'"```
it shows a running icon in the doc, if I try to show all windows it says "No Available Windows".
you get initial focus and you can get focus by clicking on the application icon; but I do not know how to make the terminal visible again.  Seems not a z-order placement but hidden or no window at all.

AlwaysOnTop runs fine in my tests.  If the behavior is so radically different (no window)...I'd prefer calling that "Hidden" over over "AlwaysOnBottom" and making it work the same across windows and mac.  I will look at my windows machine again...

davehorner commented:
> I'd probably accept a PR sent by someone else

Let me know how you want to proceed; I will continue to merge.  If you want the additional enum or something more, I'm a willing contributor.  I'd like to see this PR go or close it out.

chrisduerr commented:
> Let me know how you want to proceed; I will continue to merge. If you want the additional enum or something more, I'm a willing contributor. I'd like to see this PR go or close it out.

Feel free to not rebase this PR, I can easily do that myself if we want to merge it.

I'm sorry about the lack of communication, but the main issue right now is that neither me nor @kchibisov are really convinced we want this feature in its current state.

It's simple enough to consider just merging, but it does just dump yet another config option into the config, which means merging should be carefully considered.

chrisduerr commented on extra/man/alacritty.5.scd:
```diff
@@ -195,6 +195,12 @@ This section documents the *[window]* table of the configuration file.
 
 	Default: _"None"_
 
+*level* = _"Normal"_ | _"AlwaysOnTop"_
+
+	Sets the window level (Normal, AlwaysOnTop).
+
+	Default: _"Normal"_
```
The different window levels should be documented, see `startup_mode` for reference.

Having the variants in parenthesis after the description is not beneficial, so that should be removed.

chrisduerr commented on CHANGELOG.md:
```diff
@@ -10,6 +10,10 @@ Notable changes to the `alacritty_terminal` crate are documented in its
 
+### Added
+
+- window.level sets preferred window level (Normal, AlwaysOnTop).
```
```suggestion
- Config option `window.level = "AlwaysOnTop"` to force Alacritty to always be the toplevel window
```

davehorner commented:
I respect the process and no worries if it doesn't go.  thanks.

chrisduerr reviewed (approved):
Let's just send it, it's pretty minimal.
